File: contextscore-cc/src/snapshot/recovery.ts
import type { ContextSnapshot } from "../core/models.js";
import { SnapshotStore } from "./store.js";

/**
 * Generates structured recovery context to inject after compaction.
 * Reads the latest snapshot and produces a focused recovery prompt.
 */
export class ContextRecovery {
  private store: SnapshotStore;

  constructor(projectRoot?: string) {
    this.store = new SnapshotStore(projectRoot);
  }

  /**
   * Generate recovery text to inject after compaction.
   * Returns null if no snapshot is available.
   */
  recover(sessionId?: string): string | null {
    const snapshot = this.store.loadLatest(sessionId);
    if (!snapshot) return null;
    return this.formatRecovery(snapshot);
  }

  private formatRecovery(snap: ContextSnapshot): string {
    const lines: string[] = [
      "═══ CONTEXT RECOVERY (post-compaction) ═══",
      "",
      `Session: ${snap.sessionId}`,
      `Snapshot taken: ${snap.timestamp}`,
      `Quality score at snapshot: ${snap.qualityScore}/100`,
      `Tokens at snapshot: ${snap.tokenCount.toLocaleString()}`,
      "",
    ];

    // Current task
    if (snap.currentTask) {
      lines.push("## Current Task");
      lines.push(snap.currentTask);
      lines.push("");
    }

    // Active files
    if (snap.activeFiles.length) {
      lines.push("## Active Files");
      for (const f of snap.activeFiles.slice(0, 15)) {
        lines.push(`  - ${f}`);
      }
      lines.push("");
    }

    // Decisions (most critical for continuity)
    if (snap.decisions.length) {
      lines.push("## Key Decisions Made (DO NOT reverse these)");
      for (const d of snap.decisions.slice(0, 10)) {
        lines.push(`  - ${d.description}`);
        if (d.affectedFiles.length) {
          lines.push(`    Files: ${d.affectedFiles.join(", ")}`);
        }
      }
      lines.push("");
    }

    // Entities
    const fileEntities = snap.entities.filter(e => e.type === "file");
    const codeEntities = snap.entities.filter(e => e.type === "class" || e.type === "config");
    if (codeEntities.length) {
      lines.push("## Key Code Entities");
      for (const e of codeEntities.slice(0, 15)) {
        lines.push(`  - ${e.name} (${e.type})`);
      }
      lines.push("");
    }

    // Patterns
    if (snap.patterns.length) {
      lines.push("## Established Patterns & Conventions");
      for (const p of snap.patterns.slice(0, 8)) {
        lines.push(`  - ${p}`);
      }
      lines.push("");
    }

    // Error resolutions
    if (snap.errorResolutions.length) {
      lines.push("## Resolved Errors (DO NOT re-introduce)");
      for (const e of snap.errorResolutions.slice(0, 5)) {
        lines.push(`  - ${e}`);
      }
      lines.push("");
    }

    lines.push("═══ END CONTEXT RECOVERY ═══");
    return lines.join("\n");
  }
}

File: contextscore/src/contextscore/snapshot/recovery.py
"""
Context Recovery — generates structured recovery context to inject after compaction.

Reads the latest snapshot and produces a focused recovery prompt.
Ported from contextscore-cc/src/snapshot/recovery.ts.
"""

from __future__ import annotations

from pathlib import Path
from typing import Optional

from contextscore.models import ContextSnapshot
from contextscore.snapshot.store import SnapshotStore


class ContextRecovery:
    """Generates structured recovery context to inject after compaction."""

    def __init__(self, project_root: str | Path | None = None) -> None:
        self._store = SnapshotStore(project_root)

    def recover(self, session_id: Optional[str] = None) -> Optional[str]:
        """
        Generate recovery text to inject after compaction.

        Args:
            session_id: Optional session ID to filter snapshots.

        Returns:
            Formatted recovery text, or None if no snapshot is available.
        """
        snapshot = self._store.load_latest(session_id)
        if snapshot is None:
            return None
        return self._format_recovery(snapshot)

    def _format_recovery(self, snap: ContextSnapshot) -> str:
        lines: list[str] = [
            "═══ CONTEXT RECOVERY (post-compaction) ═══",
            "",
            f"Session: {snap.session_id}",
            f"Snapshot taken: {snap.timestamp}",
            f"Quality score at snapshot: {snap.quality_score}/100",
            f"Tokens at snapshot: {snap.token_count:,}",
            "",
        ]

        # Current task
        if snap.current_task:
            lines.append("## Current Task")
            lines.append(snap.current_task)
            lines.append("")

        # Active files
        if snap.active_files:
            lines.append("## Active Files")
            for f in snap.active_files[:15]:
                lines.append(f"  - {f}")
            lines.append("")

        # Decisions (most critical for continuity)
        if snap.decisions:
            lines.append("## Key Decisions Made (DO NOT reverse these)")
            for d in snap.decisions[:10]:
                lines.append(f"  - {d.description}")
                if d.affected_files:
                    lines.append(f"    Files: {', '.join(d.affected_files)}")
            lines.append("")

        # Entities
        code_entities = [e for e in snap.entities if e.type in ("class", "config")]
        if code_entities:
            lines.append("## Key Code Entities")
            for e in code_entities[:15]:
                lines.append(f"  - {e.name} ({e.type})")
            lines.append("")

        # Patterns
        if snap.patterns:
            lines.append("## Established Patterns & Conventions")
            for p in snap.patterns[:8]:
                lines.append(f"  - {p}")
            lines.append("")

        # Error resolutions
        if snap.error_resolutions:
            lines.append("## Resolved Errors (DO NOT re-introduce)")
            for e in snap.error_resolutions[:5]:
                lines.append(f"  - {e}")
            lines.append("")

        lines.append("═══ END CONTEXT RECOVERY ═══")
        return "\n".join(lines)
